src/orders/views/OrderReturn/utils.tsx

```diff
@@ -0,0 +1,66 @@
+import { OrderRefundAmountCalculationMode } from "@saleor/orders/components/OrderRefundPage/form";
+import {
+  FormsetQuantityData,
+  FormsetReplacementData,
+  OrderReturnFormData
+} from "@saleor/orders/components/OrderReturnPage/form";
+import { getById } from "@saleor/orders/components/OrderReturnPage/utils";
+import {
+  OrderReturnFulfillmentLineInput,
+  OrderReturnLineInput,
+  OrderReturnProductsInput
+} from "@saleor/types/globalTypes";
+
+export const getParsedData = ({
+  amount,
+  amountCalculationMode,
+  refundShipmentCosts,
+  fulfiledItemsQuantities,
+  unfulfiledItemsQuantities,
+  itemsToBeReplaced
+}: OrderReturnFormData): OrderReturnProductsInput => {
+  const shouldRefund = !!amount;
+  const amountToRefund =
+    amountCalculationMode === OrderRefundAmountCalculationMode.MANUAL
+      ? amount
+      : undefined;
+
+  const orderLines = getParsedLineData<OrderReturnLineInput>(
+    unfulfiledItemsQuantities,
+    itemsToBeReplaced,
+    "orderLineId"
+  );
+
+  const fulfillmentLines = getParsedLineData<OrderReturnFulfillmentLineInput>(
+    fulfiledItemsQuantities,
+    itemsToBeReplaced,
+    "fulfillmentLineId"
+  );
+
+  return {
+    amountToRefund,
+    fulfillmentLines,
+    includeShippingCosts: refundShipmentCosts,
+    orderLines,
+    refund: shouldRefund
+  };
+};
+
+function getParsedLineData<T>(
+  itemsQuantities: FormsetQuantityData,
+  itemsToBeReplaced: FormsetReplacementData,
+  idKey: "fulfillmentLineId" | "orderLineId"
+): T[] {
+  return itemsQuantities.reduce((result, { value: quantity, id }) => {
+    if (!quantity) {
+      return result;
+    }
+
+    const shouldReplace = !!itemsToBeReplaced.find(getById(id))?.value;
+
+    return [
+      ...result,
+      ({ [idKey]: id, quantity, replace: shouldReplace } as unknown) as T
```

Comment:
Why is it treated as `unknown`? Returned objects have strictly defined type (has `[idKey]`, `quantity` and `replace` keys), while `unknown` allows to pass arbitrary generic T type what may cause that returned value is not consistent with T type. In other words typescript will not warn if you write `getParsedLineData<{ anything: any }>`, while it should, as `anything` doesn't exist in returned objects.

Maybe restricting allowed types to `getParsedLineData<T extends OrderReturnLineInput | OrderReturnFulfillmentLineInput>` may help?

Reply:
Good point! Unfortunately the `uknown` conversion was still needed by ts, but indeed now when initialing `getParsedLineData<{foo: boolean}>` typescript throws an error.